crates/table/src/layout.rs

```diff
@@ -440,19 +440,96 @@ impl AlgebraicTypeLayout {
     pub fn algebraic_type(&self) -> AlgebraicType {
         match self {
             AlgebraicTypeLayout::Primitive(prim) => prim.algebraic_type(),
-            AlgebraicTypeLayout::VarLen(var_len) => var_len.algebraic_type(),
+            AlgebraicTypeLayout::VarLen(var_len) => var_len.algebraic_type().clone(),
             AlgebraicTypeLayout::Product(prod) => AlgebraicType::Product(prod.product_type()),
             AlgebraicTypeLayout::Sum(sum) => AlgebraicType::Sum(sum.sum_type()),
         }
     }
+
+    /// Check that this `AlgebraicTypeLayout` is a sub-layout of `other`.
+    ///
+    /// A layout L is a sub-layout of L' if any value deserialized to `AlgebraicValue` via L
+    /// can also be deserialized to an identical `AlgebraicValue` via L'. We additionally require
+    /// that the names in `L` and `L'` match.
+    ///
+    /// This means that:
+    ///
+    /// - `self.size() == other.size()`
+    ///
+    /// - `self.align() >= other.align()` (in other words, `other.align()` divides `self.align()`)
+    ///
+    /// - If `self` is a product type, then:
+    ///     - `other` must also be a product type.
+    ///     - `other` must have fields of the same names in the same positions,
+    ///       with each field's layout being a sub-layout of the corresponding field in `other`.
+    ///
+    /// - If `self` is a sum type, then:
+    ///     - `other` must also be a sum type,
+    ///     - `self`'s variant names must be a *prefix* of `other`'s variant names.
+    ///     - `self`'s variant layouts in the prefix must be sub-layouts of the corresponding variant layouts in `other`.
+    ///     - The additional layouts in `other` must not have alignment requirements larger than the alignment of `self`.
```

Comment:
This is, or should be, implied by the above requirement that `self.align() >= other.align()`, right?